File: partiql-plan/src/main/java/org/partiql/plan/rex/RexPathKey.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;

import java.util.List;

/**
 * Logical path by key lookup expression abstract base class.
 */
public abstract class RexPathKey extends RexBase {

    /**
     * @return new RexPathKey instance
     */
    @NotNull
    public static RexPathKey create(@NotNull Rex operand, @NotNull Rex key) {
        return new Impl(operand, key);
    }

    /**
     * @return operand rex (operand 0)
     */
    @NotNull
    public abstract Rex getOperand();

    /**
     * @return key rex.
     */
    @NotNull
    public abstract Rex getKey();

    @Override
    @NotNull
    protected final RexType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected List<Operand> operands() {
        Operand c0 = Operand.single(getOperand());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitPathKey(this, ctx);
    }

    private static class Impl extends RexPathKey {

        private final Rex operand;
        private final Rex key;

        private Impl(@NotNull Rex operand, @NotNull Rex key) {
            this.operand = operand;
            this.key = key;
        }

        @NotNull
        @Override
        public Rex getOperand() {
            return operand;
        }

        @NotNull
        @Override
        public Rex getKey() {
            return key;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rex/RexPathKey.kt
package org.partiql.plan.rex

/**
 * Logical operator for path lookup by key.
 */
public interface RexPathKey : Rex {

    public fun getOperand(): Rex

    public fun getKey(): Rex

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitPathKey(this, ctx)
}

/**
 * Standard internal implementation for [RexPathKey].
 */
internal class RexPathKeyImpl(operand: Rex, key: Rex, type: RexType) : RexPathKey {

    // DO NOT USE FINAL
    private var _operand = operand
    private var _key = key
    private var _type = type

    override fun getOperand() = _operand

    override fun getKey() = _key

    override fun getType(): RexType = _type

    override fun getChildren(): Collection<Rex> = listOf(_operand, _key)
}
